### PR TITLE
NDRS-519 AuctionInfo rpc endpoint

### PR DESCRIPTION
**Pull request by EdHastingsCasperLabs**
This PR adds a new rpc endpoint to api-server component, which reports current bids and validators tracked in global state.

```
curl -i -X POST -H 'Content-Type: application/json' -d '{"id":1,"jsonrpc":"2.0","method":"state_get_auction_info", "params":{}  }' 'http://localhost:<port>/rpc'
```

sample output:
```
{
  "state_root_hash": "83ff5991c599f1a038deeeb4eebc61c297a48c41814445320bcb4b4b229c0f62",
  "era_id": 1,
  "validator_weights": {
    "0248..6873": "100000000000001",
    "449c..0d31": "300000000000000",
    "8f5a..7509": "400000000000000",
    "b3fe..0d36": "200000000000000",
    "f60b..0c18": "500000000000000"
  },
  "bids": {
    "0248..6873": {
      "bonding_purse": "uref-66a5ef6413533dde84b25e8e835269f5ba9605ebd63e41e32bfce1f96260ad12-007",
      "staked_amount": "100000000000001",
      "delegation_rate": 0,
      "funds_locked": 15
    },
    "449c..0d31": {
      "bonding_purse": "uref-21c08087523473ff98d3a5fb158f48766b805e38c5507fbd825325bd9d71060c-007",
      "staked_amount": "300000000000000",
      "delegation_rate": 0,
      "funds_locked": 15
    },
    "8f5a..7509": {
      "bonding_purse": "uref-ec40e49c6a67b59ba0af3c68111dc232ecf20e860f03d28916dce87702bdfe28-007",
      "staked_amount": "400000000000000",
      "delegation_rate": 0,
      "funds_locked": 15
    },
    "b3fe..0d36": {
      "bonding_purse": "uref-8ea53a50051c12e92df7f3984941fa09fa7cf6efa58c43734cde97098a96edb1-007",
      "staked_amount": "200000000000000",
      "delegation_rate": 0,
      "funds_locked": 15
    },
    "f60b..0c18": {
      "bonding_purse": "uref-6631919b713c296255116450bd915e75a2b158b1a1ae0cf5d8b1ea8ca507eaa6-007",
      "staked_amount": "500000000000000",
      "delegation_rate": 0,
      "funds_locked": 15
    }
  }
}
```

### CODE REVIEW

File: node/src/components/api_server/rpcs/state.rs [hunk 1]
```diff
@@ -212,3 +222,138 @@ impl RpcWithParamsExt for GetBalance {
...
+                )
+                .await;
+
+            let validator_weights = {
+                if let Ok(validator_weights) = era_validators_result {
+                    validator_weights
+                } else {
+                    None
+                }
+            };
```
Comment (Fraser999) [inline node/src/components/api_server/rpcs/state.rs:350]:
```suggestion
            let validator_weights = era_validators_result.ok().flatten();
```

File: node/src/components/api_server/rpcs/state.rs [hunk 2]
```diff
@@ -212,3 +222,138 @@ impl RpcWithParamsExt for GetBalance {
...
+
+            let auction_state =
+                AuctionState::new(global_state_hash, era_id, bids, validator_weights);
+            info!("AuctionState --auction_state: {:?}", auction_state);
```
Comment (Fraser999) [inline node/src/components/api_server/rpcs/state.rs]: Was this left in by mistake?

File: node/src/effect/requests.rs [hunk 3]
```diff
@@ -435,6 +456,15 @@ impl<I> Display for ApiRequest<I> {
                 "query {}, base_key: {}, path: {:?}",
                 state_root_hash, base_key, path
             ),
+            ApiRequest::QueryEraValidators {
+                global_state_hash,
+                era_id,
+                ..
+            } => write!(
+                formatter,
+                "auction {}, era_id: {}",
```
Comment (Fraser999) [inline node/src/effect/requests.rs]:
```suggestion
                "global_state_hash {}, era_id {}",
```

File: node/src/types/json_compatibility/auction_state.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+pub type ValidatorWeights = BTreeMap<json_compatibility::PublicKey, U512>;
+
+/// An entry in a founding validator map.
+#[derive(PartialEq, Debug, Deserialize, Serialize, Clone)]
+pub struct Bid {
+    /// The purse that was used for bonding.
+    pub bonding_purse: String,
+    /// The total amount of staked tokens.
+    pub staked_amount: U512,
+    /// Delegation rate
```
Comment (Fraser999) [inline node/src/types/json_compatibility/auction_state.rs:25]:
```suggestion
    /// Delegation rate.
```

File: node/src/types/json_compatibility/auction_state.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            funds_locked: bid.funds_locked,
+        }
+    }
+}
+
+/// Data structure summarizing auction contract data.
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct AuctionState {
+    /// Global state hash
+    pub global_state_hash: Digest,
```
Comment (Fraser999) [inline node/src/types/json_compatibility/auction_state.rs:49]: This is renamed in all user-facing pieces to `state_root_hash`.  I think that we should rename it here too.

File: node/src/components/api_server/rpcs/state.rs [hunk 6]
```diff
@@ -212,3 +222,137 @@ impl RpcWithParamsExt for GetBalance {
...
+            let validator_weights = {
+                if let Ok(validator_weights) = era_validators_result {
+                    validator_weights
+                } else {
+                    None
+                }
+            };
+
+            let auction_state = AuctionState::new(state_root_hash, era_id, bids, validator_weights);
+            debug!("AuctionState --auction_state: {:?}", auction_state);
```
Comment (Fraser999) [inline node/src/components/api_server/rpcs/state.rs:353]:
```suggestion
            debug!("responding to client with {:?}", auction_state);
```

File: node/src/types/json_compatibility/auction_state.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                    let mut ret: BTreeMap<json_compatibility::PublicKey, Bid> = BTreeMap::new();
+                    for item in items {
+                        let key = json_compatibility::PublicKey::from(item.0);
+                        let value = item.1.into();
+                        ret.insert(key, value);
+                    }
+                    Some(ret)
+                }
+            }
+        };
```
Comment (Fraser999) [inline node/src/types/json_compatibility/auction_state.rs:79]:
```suggestion
        let bids = bids.map(|items| {
            items
                .into_iter()
                .map(|(public_key, bid)| (public_key.into(), bid.into()))
                .collect()
        });
```

File: node/src/types/json_compatibility/auction_state.rs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                    let mut ret: BTreeMap<json_compatibility::PublicKey, U512> = BTreeMap::new();
+                    for item in items {
+                        let key = json_compatibility::PublicKey::from(item.0);
+                        let value = item.1;
+                        ret.insert(key, value);
+                    }
+                    Some(ret)
+                }
+            }
+        };
```
Comment (Fraser999) [inline node/src/types/json_compatibility/auction_state.rs:94]:
```suggestion
        let validator_weights = validator_weights.map(|items| {
            items
                .into_iter()
                .map(|(public_key, weight)| (public_key.into(), weight))
                .collect()
        });
```

File: node/src/types/json_compatibility/public_key.rs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+/// Stringified PublicKey encoding.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct PublicKey(String);
+
+impl From<casper_types::PublicKey> for PublicKey {
+    fn from(public_key: casper_types::PublicKey) -> Self {
+        match public_key {
+            casper_types::PublicKey::Ed25519(inner) => PublicKey(format!("{:10}", HexFmt(inner))),
+            casper_types::PublicKey::Secp256k1(inner) => PublicKey(format!("{:10}", HexFmt(inner))),
```
Comment (Fraser999) [inline node/src/types/json_compatibility/public_key.rs:12]: Will this abbreviated form be sufficient?  If a user wants to copy-paste the returned public key in order to enter it into a subsequent client command, they won't be able to do that.

That may be ok here though?

File: types/src/access_rights.rs [hunk 10]
```diff
@@ -11,6 +12,7 @@ bitflags! {
     /// A struct which behaves like a set of bitflags to define access rights associated with a
     /// [`URef`](crate::URef).
     #[allow(clippy::derive_hash_xor_eq)]
+    #[derive(Deserialize, Serialize)]
```
Comment (Fraser999) [inline types/src/access_rights.rs:15]: We don't need this change do we?

File: node/src/components/api_server/rpcs/state.rs (hunk 2 above)
Comment (Fraser999) [reply node/src/components/api_server/rpcs/state.rs]: Never mind - I added that comment before your second commit.  I refreshed and my comment seemed to have disappeared, but apparently not :man_shrugging:

File: node/src/effect/requests.rs (hunk 3 above)
Comment (EdHastingsCasperLabs) [reply node/src/effect/requests.rs]: this is following the pattern established by the preexisting query and balance variants.

```
ApiRequest::QueryGlobalState {
                state_root_hash,
                base_key,
                path,
                ..
            } => write!(
                formatter,
                "query {}, base_key: {}, path: {:?}",
                state_root_hash, base_key, path
            ),
            ApiRequest::QueryEraValidators {
                state_root_hash,
                era_id,
                ..
            } => write!(formatter, "auction {}, era_id: {}", state_root_hash, era_id),
            ApiRequest::GetBalance {
                state_root_hash,
                purse_uref,
                ..
            } => write!(
                formatter,
                "balance {}, purse_uref: {}",
                state_root_hash, purse_uref
            ),
```

File: node/src/types/json_compatibility/auction_state.rs (hunk 5 above)
Comment (EdHastingsCasperLabs) [reply node/src/types/json_compatibility/auction_state.rs:49]: Already did.

File: node/src/types/json_compatibility/public_key.rs (hunk 9 above)
Comment (EdHastingsCasperLabs) [reply node/src/types/json_compatibility/public_key.rs:12]: There's no subsequent command that I'm aware of that would require this.

File: types/src/access_rights.rs (hunk 10 above)
Comment (EdHastingsCasperLabs) [reply types/src/access_rights.rs:15]: Ah, that was to allow serialization of Bid (bonding_purse:URef), but ended up creating a json_compatibility struct for Bid instead. Will remove.

File: node/src/components/api_server/rpcs/state.rs (hunk 1 above)
Comment (EdHastingsCasperLabs) [reply node/src/components/api_server/rpcs/state.rs]: Did this.

File: node/src/types/json_compatibility/auction_state.rs (hunk 4 above)
Comment (EdHastingsCasperLabs) [reply node/src/types/json_compatibility/auction_state.rs]: Did this.

File: node/src/components/api_server/rpcs/state.rs (hunk 6 above)
Comment (EdHastingsCasperLabs) [reply node/src/components/api_server/rpcs/state.rs]: Did this.

File: node/src/types/json_compatibility/auction_state.rs (hunk 7 above)
Comment (EdHastingsCasperLabs) [reply node/src/types/json_compatibility/auction_state.rs]: Did this.

File: node/src/types/json_compatibility/auction_state.rs (hunk 8 above)
Comment (EdHastingsCasperLabs) [reply node/src/types/json_compatibility/auction_state.rs]: Did this.

File: types/src/access_rights.rs (hunk 10 above)
Comment (EdHastingsCasperLabs) [reply types/src/access_rights.rs]: Did this.

File: node/src/effect/requests.rs [hunk 11]
```diff
@@ -382,6 +382,24 @@ pub enum ApiRequest<I> {
         responder: Responder<Result<QueryResult, engine_state::Error>>,
     },
     /// Query the global state at the given root hash.
+    QueryEraValidators {
+        /// The global state hash.
```
Comment (dwerner) [inline node/src/effect/requests.rs:386]: Comment out of date with field name.

**Review by dwerner — APPROVED** [review]
One nitpick, looks good to me otherwise.

File: node/src/effect/requests.rs (hunk 11 above)
Comment (EdHastingsCasperLabs) [reply node/src/effect/requests.rs:386]: The comment is correct and explanatory; `state_root_hash` is a global state hash.

**Review by Fraser999 — APPROVED** [review]

**Pull request merged**